quickwit-indexing/src/source/push_api_source.rs

```diff
@@ -0,0 +1,355 @@
+    push_api_mailbox: &Mailbox<PushApiService>,
+    index_id: String,
+) -> anyhow::Result<()> {
+    let queue_exists_req = QueueExistsRequest {
+        queue_id: index_id.clone(),
+    };
+    if push_api_mailbox
+        .ask_for_res(queue_exists_req)
+        .await
+        .map_err(|err| anyhow::anyhow!(err.to_string()))?
+    {
+        return Ok(());
+    }
+
+    let create_queue_req = CreateQueueRequest { queue_id: index_id };
+    push_api_mailbox
+        .ask_for_res(create_queue_req)
+        .await
+        .map_err(|err| anyhow::anyhow!(err.to_string()))
+}
+
+#[async_trait]
+impl Source for PushApiSource {
+    async fn emit_batches(
+        &mut self,
+        batch_sink: &Mailbox<Indexer>,
+        ctx: &SourceContext,
+    ) -> Result<Option<Duration>, ActorExitStatus> {
+        let start_after = if self.counters.current_offset == 0 {
+            None
+        } else {
+            Some(self.counters.current_offset)
+        };
+
+        let fetch_req = FetchRequest {
+            index_id: self.params.index_id.clone(),
+            start_after,
+            num_bytes_limit: self
+                .params
+                .batch_num_bytes_threshold
+                .or(Some(BATCH_NUM_BYTES_THRESHOLD)),
+        };
+        let fetch_resp = self
+            .push_api_mailbox
+            .ask_for_res(fetch_req)
+            .await
+            .map_err(|err| anyhow::anyhow!(err.to_string()))?;
+
+        let FetchResponse {
+            first_position: first_position_opt,
+            doc_batch: doc_batch_opt,
+        } = fetch_resp;
+
+        if first_position_opt.is_none() {
+            // TODO: We are currently waiting for a constant time.
+            // Think of better way, maybe increment this (i.e wait longer) as time
+            // goes on without receiving docs.
+            return Ok(Some(Duration::from_secs(1)));
+        }
+
+        let doc_batch = doc_batch_opt.unwrap();
+        let first_position = first_position_opt.unwrap();
+        let docs = iter_doc_payloads(&doc_batch)
+            .map(|buff| String::from_utf8_lossy(buff).to_string())
+            .collect::<Vec<_>>();
+
+        if !docs.is_empty() {
+            self.counters.num_docs_processed += docs.len() as u64;
+            self.counters.current_offset = first_position + docs.len() as u64 - 1;
+
+            let mut checkpoint_delta = CheckpointDelta::default();
+            let partition_id = PartitionId::from(self.params.index_id.as_str());
+            checkpoint_delta
+                .record_partition_delta(
+                    partition_id,
+                    Position::from(self.counters.previous_offset),
+                    Position::from(self.counters.current_offset),
+                )
+                .unwrap();
+
+            let raw_doc_batch = RawDocBatch {
+                docs,
+                checkpoint_delta,
+            };
+            self.counters.previous_offset = self.counters.current_offset;
```

Comment:
Can we group the logic updating the counters in a separate method?

Reply:
Maybe I am missing your point here. a portion of that update needs to be done before saving the Checkpoint. That is two lines above and this line. Should we then move all the checkpointing stuff with counters update into another method?